LevelComponents/EntitySet.cpp

```diff
@@ -59,11 +78,173 @@ namespace LevelComponents
+        for (int i = 0; i < TilesDefaultNum; ++i)
+        {
+            tile8x8array[i] = blankTile;
+        }
+    }
+
+    /// <summary>
+    /// Re-Initialize palettes and delete the old ones.
+    /// </summary>
+    void EntitySet::ResetPalettes()
+    {
+        for (int i = 0; i < 3; ++i)
+        {
+            palettes[i].clear();
+            for (int j = 0; j < 16; ++j)
+                palettes[i].push_back(QColor(0, 0, 0, 0xFF).rgba());
+        }
+        for (int i = 3; i < 15; ++i)
+        {
+            palettes[i].clear();
+        }
+        for (int i = 3; i < 8; ++i) // load universal palettes [3, 7]
+        {
+            palettes[i] << ROMUtils::entities[6]->GetPalette(i - 3);
+        }
+        int offset = 8;
+        int localEntityId = 0;
+        bool overwriteBoxtiles = false;
+        do
+        {
+            int tmpEntityGlobalId = EntityinfoTable[localEntityId].Global_EntityID;
+            int tmpEntityPalOffset = EntityinfoTable[localEntityId].paletteOffset;
+            ++localEntityId;
+
+            int tmpEntityPalNum = ROMUtils::entities[tmpEntityGlobalId]->GetPalNum();
+            offset = tmpEntityPalNum + tmpEntityPalOffset;
+            if (offset > 14)
+                overwriteBoxtiles = true;
+            if (offset > 15)
+            {
+                // TODO: deal with exception
+                continue;
+            }
+            for (int i = tmpEntityPalOffset; i < offset; ++i) // load specified sprites' tiles
+            {
+                palettes[i].clear(); // sometimes palettes will overwrite each other
+                palettes[i] << ROMUtils::entities[tmpEntityGlobalId]->GetPalette(i - tmpEntityPalOffset);
+            }
+        } while(EntityinfoTable.size() > localEntityId);
+        if(!overwriteBoxtiles)
+        {
+            palettes[15] << ROMUtils::entities[0]->GetPalette(0);
+        }
+        for (int i = 3; i < 16; ++i)
+        {
+            if (!palettes[i].size())
+            {
+                for (int j = 0; j < 16; ++j)
+                    palettes[i].push_back(QColor(0, 0, 0, 0xFF).rgba());
+            }
+        }
+    }
+
+    /// <summary>
+    /// Re-Initialize tile8x8array and delete the old ones.
+    /// </summary>
+    void EntitySet::ResetTile8x8Array()
+    {
+        // Clean up and re-initialize the 8x8 tiles and set all the tiles to blank tiles
+        for (int i = (0x20 * 4); i < TilesDefaultNum; ++i)
+        {
+            if (tile8x8array[i] != blankTile)
+            {
+                delete tile8x8array[i];
+            }
+        }
+        // Load universal sprites
+        QVector<Tile8x8 *> tmptilesarray = ROMUtils::entities[6]->GetSpriteTiles(palettes);
+        for (int i = (0x20 * 4); i < (0x20 * 16); ++i)
+        {
+            tile8x8array[i] = tmptilesarray[i - 0x20 * 4];
+            tile8x8array[i]->SetIndex(i);
+        }
+        int offset = 8;
+        int localEntityId = 0;
+        bool overwriteBoxtiles = false;
```

Comment:
Can you add a comment about what this is used for

Reply:
done